src/ui/features/onboarding/initialization.ts

```diff
@@ -0,0 +1,33 @@
+import { getCurrentUser } from 'src/shared/getCurrentUser';
+import { openOnboarding } from 'src/shared/openOnboarding';
+import { WindowContext, WindowParam } from 'src/ui/shared/WindowContext';
+import { OnboardingInterrupt } from './errors';
+
+async function getMode(windowContext: WindowContext) {
+  const hasOnboardingUrl = document.location.hash.startsWith('#/onboarding');
+  const hasExistingUser = Boolean(await getCurrentUser());
+  return hasOnboardingUrl || (!windowContext.isPopup() && !hasExistingUser)
+    ? 'onboarding'
+    : 'wallet';
+}
+
+export async function maybeOpenOboarding() {
+  const windowContext = WindowContext.get();
+  const hasExistingUser = Boolean(await getCurrentUser());
+
+  if (windowContext.isPopup() && !hasExistingUser) {
+    const url = new URL('../../popup.html', import.meta.url);
+    openOnboarding(url);
+    throw new OnboardingInterrupt();
+  }
+  const mode = await getMode(windowContext);
+  if (mode === 'onboarding' && !windowContext.isPopup()) {
+    // TODO: setting "mode=onboarding" is duplicated here and in {openOnboarding(url)}
+    const searchParams = new URLSearchParams(window.location.search);
+    if (!searchParams.has(WindowParam.mode)) {
+      searchParams.append(WindowParam.mode, 'onboarding');
+      window.location.search = searchParams.toString();
+    }
+  }
+  return { mode } as const;
+}
```

Comment:
I renamed `userHasWallets` to `hasExistingUser` since that’s what it represents.
Honestly, I find this logic overcomplicated, but I haven’t spent enough time to debug and think about the best way to refactor it. At least it should be possible to get rid of the `OnboardingInterrupt` exception